### ISSUE TITLE
Enhancement: [no-unsafe-type-assertion] Add option to allow from unknown

### ISSUE THREAD

**Issue by KDean-Dolphin** [typescript-eslint/typescript-eslint#10402]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unsafe-type-assertion/

### Description

Per the TypeScript documentation:

> The `unknown` type represents _any_ value. This is similar to the `any` type, but is safer because it’s not legal to do anything with an `unknown` value.

This means that any casting of a variable of type `unknown` is deliberate and the result of (hopefully) very careful consideration on the part of the programmer. While the `no-unsafe-type-assertion` is useful in most circumstances, I have yet to encounter any of my `unknown` code that could be rewritten as a result of this error; I ended up in `unknown` territory for a reason. I would like the option to allow casting from `unknown`.

### Fail

```TypeScript
class A {
  a: string;

  constructor(a: string) {
    this.a = a;
  }
}

class B extends A {
  b: string;

  constructor(a: string, b: string) {
    super(a);
    this.b = b;
  }
}

function shouldUseInstanceOf(a: A) {
  if (someExternalConditionSatisfied) {
    // Should, and does, error here.
    console.log((a as B).b);
  }
}
```

### Pass

```TypeScript
class A {
  a: string;

  constructor(a: string) {
    this.a = a;
  }
}

class B extends A {
  b: string;

  constructor(a: string, b: string) {
    super(a);
    this.b = b;
  }
}

function shouldUseInstanceOf(x: unknown) {
  if (someExternalConditionSatisfied) {
    // Should not error here with "ignoreUnknown" option set.
    console.log((x as B).b);
  }
}
```

**Comment by kirkwaiblinger** [comment]
Hi @KDean-Dolphin!

On its face, I'm not personally terribly swayed by this proposed option, since `unknown` -> `T` is essentially _always_ unsafe. But I don't have a strongly thought through opinion on this yet, so I might still be swayed by discussion.

> ... I have yet to encounter any of my `unknown` code that could be rewritten as a result of this error; I ended up in unknown` territory for a reason. I would like the option to allow casting from `unknown`.

I'm very confused by this because your example contains an example of code that could be rewritten as a result of this error, _and it even says how in the function name_.

[playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=MYGwhgzhAECC0G8BQ1pgFzQgFwE4EsA7AcwG4kVpgB7QnXAV2G2twAoMs8jiBKRSqmwALfBAB0YaAF405VAF8kSpKEgwAQtACmAD2zbCAExjxkqAEaZ6PcpRp08TFu042SAGmhWuBEv3NULAYAB21XXnkgkTFxCxlvKKUVADMGQmZ8WixhagYQIwBVCG0ASUcwDO0AeRS2XUx0gGtCagB3QgDKfBToeugiHErgbWpejS6gqloIahBtcRBqYnq4yMpkoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolpFStPsmSJo%2BShybooAdz7QF4MAF8QmoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```ts
function shouldUseInstanceOf(x: unknown) {
  if (x instanceof B) {
    console.log(x.b);
  }
}
```

I'm not trying to play dumb here, I'm genuinely just confused by what the point here is. Could you provide some more explanation to help us understand the motivation?

**Comment by bradzacher** [comment]
I agree `unknown` is safer than any because it doesn't allow unsafe access. But it's no more safe to do `unknown as string` than it is to do `any as string`.

Strong -1 from me here.

> I have yet to encounter any of my unknown code that could be rewritten as a result of this error; I ended up in unknown territory for a reason

Right -- the point of the rule isn't to tell you to rewrite your code.
It's to flag "hey this assertion is unsafe based on the types".
As the developer you may decide "I'm okay with the type unsafety here because I believe I've done the relevant checks". And if so -- use a disable comment and add that context so future readers can understand why you're doing something that looks unsafe.

**Comment by KDean-Dolphin** [comment]
The examples are highly artificial just to show something simple. Here's a real-world example from a file that acts as a bridge between a JavaScript tool and my TypeScript application:

```typescript
        // tag comes into the TypeScript function declared as unknown as it can represent multiple unrelated types.
        // Check for DocLocation type by checking for start and end attributes.
        if (isNonNullish(tag) && "start" in tag && "end" in tag) {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion -- Tag is known to be DocLocation.
            const docLocation = tag as DocLocation;

            console.log(`${validationError.message} at line ${docLocation.start.line}, column ${docLocation.start.column}`);
        }
```

The nature of the JavaScript tool is such that I'm 100% confident that `tag` implements the `DocLocation` interface if the `start` and `end` attributes are present. There is no way to rewrite the code.

You can see another one [here](https://github.com/KDean-Dolphin/TestConditional/blob/main/src/test-conditional.ts). There's a limitation in TypeScript where it doesn't resolve conditional types in the body of a function (there's a fix but it hasn't been released yet). To work around the errors that are generated, I change the type in line 31 to `unknown` and add a cast to the return type at line 43.

There's a fundamental difference between `any` and `unknown` that makes this rule redundant when dealing with `unknown`. With an object of type `any`, I can assign it to something that is strictly typed without TypeScript complaining about it; the rule is doing what TypeScript isn't. With an object of type `unknown`, I have to be very clear about its type before assignment, which means that I've done something with the logic to ensure that it has the correct type when I make the cast.

Basically, I want to be able to say, "I'm using `unknown` in extremely limited circumstances and when casting it to something else I've done my due diligence."

**Comment by kirkwaiblinger** [comment]
Ok, understood, and this is very helpful info! I'm still not swayed yet, but I have two lines of discussion that I'd like to unpack first

1. Can [user-defined type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) help you here? A user-defined type guard _is_ effectively a type assertion, to which you get to provide as little or as much validation as you think appropriate, which often times makes it preferabe to a raw type assertion. In your example, would it make sense to write
   ```ts
   function isNonNullish<T>(x: T): x is NonNullable<T> {
     return x != null;
   }
 
   declare const tag: unknown;
 
   type DocLocation = {
     start: {
       line: number,
       column: number,
     }
     end: {
       line: number,
       column: number,
     }
   }
 
   function isDocLocation(x: unknown): x is DocLocation {
     return x != null && typeof x === 'object' && "start" in x && "end" in x;
   }
 
   // tag comes into the TypeScript function declared as unknown as it can represent multiple unrelated types.
   // Check for DocLocation type by checking for start and end attributes.
   if (isDocLocation(tag)) {
     console.log(`error at line ${tag.start.line}, column ${tag.start.column}`);
   }
   ```
   
1. In my experience in dealing with passing C++/wasm (emscripten) objects to TS, we had a specified area of the codebase that was pretty loosey-goosey because, well, it needed to be. Therefore we conditionally disabled several lint rules for _those_ files specifically, rather than the whole codebase. I wonder whether in your case it would make sense to do something similar? Either conditionally disabling on certain files, or within the body of functions, like so?
   ```ts
   function safeAbstractionOfUnsafeStuff(x: unknown) {
     /* eslint-disable @typescript-eslint/no-unsafe-type-assertion */
     const n = x as number;
     /* eslint-enable @typescript-eslint/no-unsafe-type-assertion */
   }
   ```

Interested in your thoughts!

***

Note - I'm actually more sympathetic to a rule mode that would allow `any` -> `T` assignments than `unknown` -> `T` assignments. Right now, the rule does two separate things
- Bans type assertions that would not be allowable _assignments_ according to TS
- Bans type assertions that _would_ be allowable assignments to TS, but would be reported by [no-unsafe-assignment](https://typescript-eslint.io/rules/no-unsafe-assignment/) (which largely just exists to force you to treat `any` like `unknown`)

Therefore using the rule only really makes sense IMO for people using the rest of our no-unsafe-* suite. But I can see wanting a mode to ban _only_ the type assertions that wouldn't be allowable to TS, in order to allow usage of `any` when you want it. You may be finding yourself in a scenario where `any` is sensible.

**Comment by kirkwaiblinger** [comment]
BTW - the reason I'm averse to the `unknown` heuristic is that, once you have dev tool that provides safety, it's common to lean on it, and assume it will always provide that safety. Once I have in my head that the linter can check type assertions for me, I'm more likely to slap them down, and see if the linter complains or not, assuming all is well if it doesn't. If we put a hole in the rule, especially one that can arise implicitly, like an expression having type `unknown`, users may write more unsafe code as a result of not realizing they're hitting an intentional safety hole in a different context, where they're not thinking super hard about the types.

So explicit opt-outs like `// eslint-disable-next-line -- and here's why` or syntactically overt opt-outs are much more palatable than ones based on information that only the TS compiler knows.

**Comment by Josh-Cena** [comment]
The point of this rule is to force "careful consideration on the part of the programmer". It isn't necessarily to suggest your code is "incorrect" but, in code review, that something involving subjective belief is involved. Many of our lint rules have the scenario of code readers in mind where it's either inconvenient or impossible to hover over a variable and check its type. If you have gone through the "careful consideration" process, you could explicitly say it in the form of `eslint-disable` + a comment suggesting why it's safe in this context.

**Comment by KDean-Dolphin** [comment]
I still think it's redundant, given the difference between `unknown` and `any`. The use of `unknown` is supposed to be rare, and forces the programmer to do things that they wouldn't do with `any`. I'd like an option that recognizes that, but I can live without it.

**Comment by kirkwaiblinger** [comment]
The elephant in the room here that I just realized is that the asserted expression in your example doesn't even have type `unknown`! it's not narrowed to be a value that is not nullish, and has `start` and `end` properties.

```ts
type DocLocation = {
  start: {
    line: number,
    column: number,
  }
  end: {
    line: number,
    column: number,
  }
}

declare function isNonNullish<T>(x: T): x is NonNullable<T>;
declare const tag: unknown;

// tag comes into the TypeScript function declared as unknown as it can represent multiple unrelated types.
// Check for DocLocation type by checking for start and end attributes.
if (isNonNullish(tag) && "start" in tag && "end" in tag) {
  const docLocation = tag as DocLocation;
//                    ^? const tag: {}  

  console.log(`error at line ${docLocation.start.line}, column ${docLocation.start.column}`);
}
```
the more validation you do, the safer your code will be, and the less like `unknown` your asserted value will be. So I think that allowing `unknown` -> `T` is effectively discouraging you from validating an unknown value at all... which seems like a [perverse incentive](https://en.wikipedia.org/wiki/Perverse_incentive). Therefore, I think for me that's the nail in the coffin for the proposal as it stands. But, I think we can and should come up with some useful docs additions from this discussion.

***

Also - if you're during lots class-based stuff, note that there is [an obscure JS/TS feature about instanceof overloading](https://devblogs.microsoft.com/typescript/announcing-typescript-5-3/#instanceof-narrowing-through-symbol.hasinstance). I haven't used it myself so I don't know if it's useful, but maybe it's vaguely helpful to some of your more niche use cases? 🤷‍♂️

**Comment by bradzacher** [comment]
It looks like the team has a consensus on this. We shall reject this proposal as it is unsound, does not align with the rule's purpose, and is not broadly applicable. As such I am going to close this

Please feel free to continue discussing as it looks like you're having a valuable discussion!
